app/components/modals/NewDatasetModal.tsx

```diff
@@ -0,0 +1,122 @@
+import * as React from 'react'
+
+import { NewDatasetModal } from '../../models/modals'
+
+import { connectComponentToProps } from '../../utils/connectComponentToProps'
+
+import { dismissModal } from '../../actions/ui'
+
+import { selectModal, selectSessionUsername } from '../../selections'
+
+import nameFromTitle from '../../utils/nameFromTitle'
+
+import Modal from './Modal'
+import TextInput from '../form/TextInput'
+import Error from './Error'
+import Buttons from './Buttons'
+
+interface NewDatasetProps {
+  onDismissed: () => void
+  modal: NewDatasetModal
+  username: string
+}
+
+// 136 because dataset names cannot be longer than 144 characters,
+// and we might have to prepend `dataset_` if the title begins with a number
+const TITLE_CHARACTER_LIMIT = 136
+
+export const NewDatasetComponent: React.FunctionComponent<NewDatasetProps> = (props) => {
+  const {
+    onDismissed,
+    username
+  } = props
+
+  const [title, setTitle] = React.useState('')
+  const [datasetName, setDatasetName] = React.useState('')
```

Comment:
Because these two states are only ever set in conjunction, these should be grouped into a state object. That way they update in a single repaint and we know that updating one should de-facto update the other

```suggestion
  const [formValues, setFormValues] = React.useState({ title: '', datasetName: '' })
```